### PR TITLE
WP-10450 Prevent Column Selection in Discovery

### PR DESCRIPTION
**Pull request by jayma91**
WP-10450 Prevent Column Selection in Discovery
https://varicent.atlassian.net/browse/WP-10450

changes: 
- prevent column selection in tap for unsupported type
- drop columns before querying

### CODE REVIEW

File: tap_dynamics/sync.py [hunk 1]
```diff
@@ -32,8 +35,15 @@ def sync(config, config_path, state, catalog):
...
 
     state = singer.set_currently_syncing(state, None)
     singer.write_state(state)
+
+
+def should_sync_column(metadata, field_name):
+    field_metadata = metadata.get(('properties', field_name), {})
+    return singer.should_sync_field(field_metadata.get('inclusion'),
+                                    field_metadata.get('selected'),
+                                    True)
```
Comment (jayma91) [inline tap_dynamics/sync.py:53]: source code for singer.should_sync_field:
image

**Review by ChrisLing1 — DISMISSED** [review]

**Review by JiawenHou — DISMISSED** [review]

File: tap_dynamics/client.py [hunk 2]
```diff
@@ -258,3 +259,11 @@ def build_params(orderby_key: str = 'modifiedon',
             filter_param = f'{replication_key} ge {filter_value}'
             return {"$orderby": orderby_param, "$filter": filter_param}
         return {"$orderby": orderby_param}
+
+    
+    @staticmethod
+    def build_select_params(desired_columns: list):
+        select_columns = ','.join(desired_columns)
+        if len(select_columns) > MAX_SELECT_PARAM_SIZE or len(desired_columns) == 0:
```
Comment (sunil-varicent) [inline tap_dynamics/client.py]: is there size limitation from ms dynamics side?

File: tap_dynamics/streams.py [hunk 3]
```diff
@@ -112,15 +112,16 @@ class IncrementalStream(BaseStream):
     replication_method = 'INCREMENTAL'
     batched = False
 
-    def get_records(self, max_pagesize: int = 100, bookmark_datetime: str = None):
+    def get_records(self, max_pagesize: int = 100, desired_columns: list = [], bookmark_datetime: str = None):
```
Comment (sunil-varicent) [inline tap_dynamics/streams.py]: its better to initialize list inside function. Kevin recently posted this in dev. https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments

File: tap_dynamics/streams.py [hunk 4]
```diff
@@ -183,7 +184,7 @@ class FullTableStream(BaseStream):
     replication_method = 'FULL_TABLE'
 
     # pylint: disable=arguments-differ
-    def get_records(self, max_pagesize: int = 100):
+    def get_records(self, max_pagesize: int = 100, desired_columns: list = []):
```
Comment (sunil-varicent) [inline tap_dynamics/streams.py]: same comment as above.

File: tap_dynamics/client.py (hunk 2 above)
Comment (jayma91) [reply tap_dynamics/client.py]: I couldn't find exact limitation on ms dynamics, but instead used limit on url length that most browser would accept. Most browser seems to accept 2048 chars, so left a some room for other params and url length. 
https://stackoverflow.com/questions/417142/what-is-the-maximum-length-of-a-url-in-different-browsers

Comment (sunil-varicent) [reply tap_dynamics/client.py]: Do we return error or we select everything when its greater than limit?

Comment (jayma91) [reply tap_dynamics/client.py]: When it's greater than limit, select everything and transformer filters the records.

**Review by sunil-varicent — APPROVED** [review]

**Pull request merged**